### PR TITLE
Marksman

### PR DESCRIPTION
**Pull request by ItsSammyM**

### CODE REVIEW

File: server/src/game/chat/chat_message_variant.rs [hunk 1]
```diff
@@ -161,6 +174,7 @@ pub enum ChatMessageVariant {
     VeteranAttackedVisitor,
 
     CopAttackedVisitor,
+    MarksmanAttackedVisitor,
```
Comment (Jack-Papel) [inline server/src/game/chat/chat_message_variant.rs:177]: Make sure to remove this when the other PR gets merged

File: server/src/game/role/marksman.rs [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    },
+    ShotTownie
+}
+impl MarksmanState{
+    fn no_marks(&self)->bool{
+        !matches!(
+            self,
+            MarksmanState::Marks{marks: MarksmanMarks::One{..}} |
+            MarksmanState::Marks{marks: MarksmanMarks::Two{..}}
+        )
```
Comment (Jack-Papel) [inline server/src/game/role/marksman.rs:38]: How about matches None?

File: server/src/game/role/marksman.rs [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+            MarksmanMarks::Two{a,b} => {Box::new([*a,*b])}
+        }
+    }
+}
+
+impl Default for Marksman {
+    fn default() -> Self {
+        Self { state: MarksmanState::NotLoaded }
+    }
+}
```
Comment (Jack-Papel) [inline server/src/game/role/marksman.rs:101]: You can implement default for MarksmanState (look up rust enum default macro) and then auto-generate this using the derive macro

File: server/src/game/role/marksman.rs [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+    }
+    fn marks(&self)->Box<[PlayerReference]> {
+        if let Self::Marks{marks} = self {
+            marks.marks()
+        }else{
+            Box::new([])   
+        }
+    }
+    /// This function will mark an unmarked player or un-mark a marked player
+    /// if the action is invalid, then it will do nothing
```
Comment (Jack-Papel) [inline server/src/game/role/marksman.rs:48]: This comment is wrong

File: server/src/game/role/marksman.rs [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+                actor_ref.set_role_state(game, RoleState::Marksman(self));
+            },
+            _ => {}
+        }
+
+    }
+    fn do_day_action(mut self, game: &mut Game, actor_ref: PlayerReference, target_ref: PlayerReference) {
+        self.state.toggle_mark(target_ref);
+        if self.state.marks().len() == 0 {
+            actor_ref.set_chosen_targets(game, vec![]);
```
Comment (Jack-Papel) [inline server/src/game/role/marksman.rs:142]: Seems weird. You probably don't want this

File: server/src/game/role/marksman.rs [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+            chosen_targets.is_empty()
+        ) || (
+            chosen_targets.len() == 1 &&
+            Some(target_ref) != chosen_targets.first().copied()
+        ))
+    }
+    fn can_day_target(self, game: &Game, actor_ref: PlayerReference, target_ref: PlayerReference) -> bool {
+        game.current_phase().is_night() &&
+        target_ref.alive(game) &&
+        if let MarksmanState::Marks {..} = self.state {true}else{false} &&
```
Comment (Jack-Papel) [inline server/src/game/role/marksman.rs:165]: matches!

File: server/src/game/role/marksman.rs [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+        actor_ref != target_ref
```
Comment (Jack-Papel) [inline server/src/game/role/marksman.rs:166]: And actor is alive and actor is not jailed

File: server/tests/role.rs [hunk 8]
```diff
@@ -845,6 +846,26 @@ fn mafioso_cant_kill_mafia() {
...
+fn marksman_basic() {
+    kit::scenario!(game in Night 2 where
+        mk: Marksman,
+        dt: Detective,
+        gf: Godfather
+    );
+
+    assert!(dt.set_night_target(gf));
+    assert!(mk.day_target(dt));
+    // assert!(mk.day_target(gf));
```
Comment (Jack-Papel) [inline server/tests/role.rs:859]: Remove

File: server/src/game/role/marksman.rs [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+    }
+    fn convert_targets_to_visits(self, _game: &Game, _actor_ref: PlayerReference, target_refs: Vec<PlayerReference>) -> Vec<Visit> {
+        if target_refs.len() == 2 {
+            vec![
+                Visit{ target: target_refs[0], attack: false },
```
Comment (Jack-Papel) [inline server/src/game/role/marksman.rs:171]: We should talk about attacking visits versus indirect visits.. this never really sat right with me and I also don't quite get it.
Just a reminder, no action required here

**Review by Jack-Papel — APPROVED** [review]
Address comments

File: server/src/game/role/marksman.rs (hunk 2 above)
Comment (ItsSammyM) [reply server/src/game/role/marksman.rs:38]: There are more cases i would have to handle there
No marks, not loaded, and shot townie. This works

File: server/src/game/role/marksman.rs (hunk 5 above)
Comment (ItsSammyM) [reply server/src/game/role/marksman.rs:142]: ??? Im resolving this and ignoring it

File: client/src/menu/game/gameScreenContent/PlayerListMenu.tsx [hunk 10]
```diff
@@ -266,6 +266,16 @@ export default class PlayerListMenu extends React.Component<PlayerListMenuProps,
...
+                            || 
+                            (
+                                this.state.roleState?.type === "marksman" && 
+                                this.state.roleState.state.type === "marks" &&
+                                (
+                                    (this.state.roleState.state.marks.type === "one" && this.state.roleState.state.marks.a === player.index) ||
+                                    (this.state.roleState.state.marks.type === "two" && this.state.roleState.state.marks.a === player.index) ||
+                                    (this.state.roleState.state.marks.type === "two" && this.state.roleState.state.marks.b === player.index)
+                                )
+                            )
```
[the remark this answers is not in the document]
Comment (ItsSammyM) [reply client/src/menu/game/gameScreenContent/PlayerListMenu.tsx:278]: No it cant

[the remark this answers is not in the document]
Comment (ItsSammyM) [reply client/src/menu/game/gameScreenContent/PlayerListMenu.tsx:278]: server says so

**Pull request merged**